[BUG]: Missing inclusion of math.h during compilation with clang in C++ mode.

### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Compile-time Error

### Component

libcu++

### Describe the bug

Some of the header compilation tests under libcudacxx/test/internal_headers appear to fail in C++ compilation mode because they rely on global scope declarations that require inclusion of "math.h"  E.g.

```
In file included from blaze-out/k8-fastbuild/bin/third_party/gpus/cccl/v3_1/libcudacxx/test_internal_cuda_cxx_wrappers/cuda/std/__cmath/isnan.h.cc:23:
blaze-out/k8-fastbuild/bin/third_party/gpus/cuda/_virtual_includes/_stage/third_party/gpus/cuda/include/cuda/std/__cmath/isnan.h:49:14: error: no member named 'isnan' in the global namespace
   49 |     return ::isnan(__x);
      |              ^~~~~
1 error generated.
```

What appears to happen is that inclusion of math.h is predicated on `_CCCL_CUDA_COMPILER(CLANG)`:
https://github.com/NVIDIA/cccl/blob/b937b649f8665dc277fe2db2372f0dbe75faf86e/libcudacxx/include/cuda/std/__cmath/isnan.h#L31-L33

However, during C++ compilation we end up with `#define _CCCL_CUDA_COMPILER_CLANG() _CCCL_VERSION_INVALID()` because we are compiling in C++ mode, and `__CUDA__` is not set:
https://github.com/NVIDIA/cccl/blob/b937b649f8665dc277fe2db2372f0dbe75faf86e/libcudacxx/include/cuda/std/__cccl/compiler.h#L118-L121



### How to Reproduce

The issue shows up in C++ compilation mode with clang and libc++. I can not reproduce it on godbolt, probably because CCCL there is not recent enough.

### Expected behavior

math.h should be included in C++ compilations with clang before using global scope math functions.  Perhaps it should be done universally, as it's math.h's job to provide those global scope math function declarations. The fact that CUDA SDK headers (or compiler) may include the header in CUDA compilation mode is an implementation detail that should not be relied on.

### Reproduction link

_No response_

### Operating System

_No response_

### nvidia-smi output

_No response_

### NVCC version

_No response_

--- a/libcudacxx/include/cuda/std/__cmath/error_functions.h
+++ b/libcudacxx/include/cuda/std/__cmath/error_functions.h
@@ -25,6 +25,10 @@
 #include <cuda/std/__floating_point/fp.h>
 #include <cuda/std/__type_traits/is_integral.h>
 
+#if !_CCCL_COMPILER(NVRTC)
+#  include <math.h>
+#endif // !_CCCL_COMPILER(NVRTC)
+
 #include <cuda/std/__cccl/prologue.h>
 
 _CCCL_BEGIN_NAMESPACE_CUDA_STD

--- a/libcudacxx/include/cuda/std/__cmath/exponential_functions.h
+++ b/libcudacxx/include/cuda/std/__cmath/exponential_functions.h
@@ -33,10 +33,9 @@
 #include <cuda/std/cstdint>
 #include <cuda/std/limits>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/fdim.h
+++ b/libcudacxx/include/cuda/std/__cmath/fdim.h
@@ -25,6 +25,10 @@
 #include <cuda/std/__floating_point/fp.h>
 #include <cuda/std/__type_traits/is_integral.h>
 
+#if !_CCCL_COMPILER(NVRTC)
+#  include <math.h>
+#endif // !_CCCL_COMPILER(NVRTC)
+
 #include <cuda/std/__cccl/prologue.h>
 
 _CCCL_BEGIN_NAMESPACE_CUDA_STD

--- a/libcudacxx/include/cuda/std/__cmath/fma.h
+++ b/libcudacxx/include/cuda/std/__cmath/fma.h
@@ -27,6 +27,10 @@
 #include <cuda/std/__type_traits/is_same.h>
 #include <cuda/std/__type_traits/promote.h>
 
+#if !_CCCL_COMPILER(NVRTC)
+#  include <math.h>
+#endif // !_CCCL_COMPILER(NVRTC)
+
 #include <cuda/std/__cccl/prologue.h>
 
 _CCCL_BEGIN_NAMESPACE_CUDA_STD

--- a/libcudacxx/include/cuda/std/__cmath/fpclassify.h
+++ b/libcudacxx/include/cuda/std/__cmath/fpclassify.h
@@ -29,10 +29,9 @@
 #include <cuda/std/__type_traits/is_integral.h>
 #include <cuda/std/limits>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/gamma.h
+++ b/libcudacxx/include/cuda/std/__cmath/gamma.h
@@ -27,10 +27,9 @@
 
 #include <nv/target>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/hyperbolic_functions.h
+++ b/libcudacxx/include/cuda/std/__cmath/hyperbolic_functions.h
@@ -27,10 +27,9 @@
 
 #include <nv/target>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/hypot.h
+++ b/libcudacxx/include/cuda/std/__cmath/hypot.h
@@ -34,10 +34,9 @@
 
 #include <nv/target>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/inverse_hyperbolic_functions.h
+++ b/libcudacxx/include/cuda/std/__cmath/inverse_hyperbolic_functions.h
@@ -27,10 +27,9 @@
 
 #include <nv/target>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/inverse_trigonometric_functions.h
+++ b/libcudacxx/include/cuda/std/__cmath/inverse_trigonometric_functions.h
@@ -29,10 +29,9 @@
 
 #include <nv/target>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/isfinite.h
+++ b/libcudacxx/include/cuda/std/__cmath/isfinite.h
@@ -28,10 +28,9 @@
 #include <cuda/std/__floating_point/fp.h>
 #include <cuda/std/__type_traits/is_integral.h>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/isinf.h
+++ b/libcudacxx/include/cuda/std/__cmath/isinf.h
@@ -29,10 +29,9 @@
 #include <cuda/std/__type_traits/is_integral.h>
 #include <cuda/std/limits>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/isnan.h
+++ b/libcudacxx/include/cuda/std/__cmath/isnan.h
@@ -27,10 +27,9 @@
 #include <cuda/std/__type_traits/is_floating_point.h>
 #include <cuda/std/__type_traits/is_integral.h>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_HOST_COMPILATION() || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_HOST_COMPILATION() || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/logarithms.h
+++ b/libcudacxx/include/cuda/std/__cmath/logarithms.h
@@ -34,10 +34,9 @@
 
 #include <nv/target>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/remainder.h
+++ b/libcudacxx/include/cuda/std/__cmath/remainder.h
@@ -27,6 +27,10 @@
 #include <cuda/std/__type_traits/is_same.h>
 #include <cuda/std/__type_traits/promote.h>
 
+#if !_CCCL_COMPILER(NVRTC)
+#  include <math.h>
+#endif // !_CCCL_COMPILER(NVRTC)
+
 #include <cuda/std/__cccl/prologue.h>
 
 _CCCL_BEGIN_NAMESPACE_CUDA_STD

--- a/libcudacxx/include/cuda/std/__cmath/roots.h
+++ b/libcudacxx/include/cuda/std/__cmath/roots.h
@@ -25,10 +25,9 @@
 #include <cuda/std/__type_traits/enable_if.h>
 #include <cuda/std/__type_traits/is_integral.h>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/rounding_functions.h
+++ b/libcudacxx/include/cuda/std/__cmath/rounding_functions.h
@@ -28,10 +28,9 @@
 #include <cuda/std/__type_traits/is_same.h>
 #include <cuda/std/__type_traits/promote.h>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

--- a/libcudacxx/include/cuda/std/__cmath/traits.h
+++ b/libcudacxx/include/cuda/std/__cmath/traits.h
@@ -33,6 +33,10 @@
 
 #include <nv/target>
 
+#if !_CCCL_COMPILER(NVRTC)
+#  include <math.h>
+#endif // !_CCCL_COMPILER(NVRTC)
+
 #include <cuda/std/__cccl/prologue.h>
 
 _CCCL_BEGIN_NAMESPACE_CUDA_STD

--- a/libcudacxx/include/cuda/std/__cmath/trigonometric_functions.h
+++ b/libcudacxx/include/cuda/std/__cmath/trigonometric_functions.h
@@ -28,10 +28,9 @@
 
 #include <nv/target>
 
-// MSVC and clang cuda need the host side functions included
-#if _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#if !_CCCL_COMPILER(NVRTC)
 #  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
+#endif // !_CCCL_COMPILER(NVRTC)
 
 #include <cuda/std/__cccl/prologue.h>
 

